File: multi-processor-programming-study/solutions/cpp/exercise01_hello_threads_solution.cpp
/**
 * SOLUTION for Exercise 01: Hello Threads! (C++)
 *
 * This file contains the reference implementation.
 * Try to solve the exercise yourself before looking at this!
 */

#include <iostream>
#include <thread>
#include <vector>

namespace multiprocessor::basics {

/**
 * SOLUTION: Create and return a thread that prints message
 */
std::thread create_thread(int thread_id) {
    return std::thread([thread_id]() {
        std::cout << "Hello from Thread " << thread_id << "!\n";
    });
}

/**
 * SOLUTION: Create multiple threads and start them
 */
std::vector<std::thread> create_and_start_threads(int count) {
    std::vector<std::thread> threads;
    for (int i = 0; i < count; ++i) {
        threads.push_back(create_thread(i));
    }
    return threads;
}

/**
 * SOLUTION: Wait for all threads to complete
 */
void wait_for_all_threads(std::vector<std::thread>& threads) {
    for (auto& t : threads) {
        if (t.joinable()) {
            t.join();
        }
    }
}

} // namespace multiprocessor::basics

int main() {
    using namespace multiprocessor::basics;

    std::cout << "Creating threads...\n";
    auto threads = create_and_start_threads(5);

    std::cout << "Waiting for threads to complete...\n";
    wait_for_all_threads(threads);

    std::cout << "All threads completed!\n";

    return 0;
}

File: multi-processor-programming-study/solutions/java/Exercise01_HelloThreads_Solution.java
package com.multiprocessor.basics;

/**
 * SOLUTION for Exercise 01: Hello Threads!
 *
 * This file contains the reference implementation.
 * Try to solve the exercise yourself before looking at this!
 */
public class Exercise01_HelloThreads_Solution {

    /**
     * SOLUTION: Create and return a thread that prints "Hello from Thread [id]!"
     */
    public static Thread createThread(int threadId) {
        return new Thread(() -> {
            System.out.println("Hello from Thread " + threadId + "!");
        });
    }

    /**
     * SOLUTION: Create multiple threads and start them all
     */
    public static Thread[] createAndStartThreads(int count) {
        Thread[] threads = new Thread[count];
        for (int i = 0; i < count; i++) {
            threads[i] = createThread(i);
            threads[i].start();
        }
        return threads;
    }

    /**
     * SOLUTION: Wait for all threads to complete
     */
    public static void waitForAllThreads(Thread[] threads) {
        for (Thread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.err.println("Thread interrupted: " + e.getMessage());
            }
        }
    }

    /**
     * Main method for testing
     */
    public static void main(String[] args) {
        System.out.println("Creating threads...");
        Thread[] threads = createAndStartThreads(5);

        System.out.println("Waiting for threads to complete...");
        waitForAllThreads(threads);

        System.out.println("All threads completed!");
    }
}
